### PR TITLE
Track full-path htlc-minimum-msat while routing

### PR DESCRIPTION
**Pull request by TheBlueMatt**
To review new changes to #646, I wrote an extension of the router fuzzer to test each route against the graph available. This found one non-regression which is fixed here, as well as the fuzzer extension and a fix to make the router fuzz target deterministic.

It needs a) a test, b) a long discussion in `get_route` about how our algorithm ultimately differs from Dijkstra's/A*, why that's ok, and some edge cases where other algorithms may be more optimal.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Pushed a few more fix commits (including some regressions from MPP) including tests for each bug. There may be one more, but these all stand.

File: lightning/src/routing/router.rs [hunk 1]
```diff
@@ -901,13 +908,23 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
-		// because we deterministically terminate the search due to low liquidity.
+		// Stop either when the recommended value is reached or if no new path was found in this
+		// iteration.
+		// In the latter case, making another path finding attempt won't help,
+		// because we deterministically terminated the search due to low liquidity.
 		if already_collected_value_msat >= recommended_value_msat || !found_new_path {
 			break 'paths_collection;
 		}
+		// Further, if this was our first walk of the graph, and we weren't limited by an
+		// htlc_minim_msat, return immediately because this path should suffice. If we were limited
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:927]: typo: htlc_minim_msat

File: lightning/src/routing/router.rs [hunk 2]
```diff
@@ -901,13 +908,23 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+		// by an htlc_minimum_msat value, find another path with a higher value, potentially
+		// allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
+		if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:930]: nit: this is mutually exclusive with the previous if because of `found_new_path`, I think it would be easier to understand their relation if merged via `else`

File: lightning/src/routing/router.rs [hunk 3]
```diff
@@ -3371,6 +3388,77 @@ mod tests {
...
+			cltv_expiry_delta: (13 << 8) | 2,
+			htlc_minimum_msat: 0,
+			htlc_maximum_msat: OptionalField::Absent,
+			fee_base_msat: 0,
+			fee_proportional_millionths: 0,
+			excess_data: Vec::new()
+		});
+
+		{
+			// Now, attempt to route 125 sats (just a bit below the capacity of 3 channels).
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:3599]: this comment says 125, but you send 90 :)

Comment (naumenkogs) [reply lightning/src/routing/router.rs:3599]: And the next line is outdated too.

File: lightning/src/routing/router.rs [hunk 4]
```diff
@@ -778,17 +790,17 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 			if have_hop_src_in_graph {
 				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
 				// really sucks, cause we're gonna need that eventually.
-				let last_hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+				let last_path_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:789]: I still don't understand this change.

File: lightning/src/routing/router.rs [hunk 5]
```diff
@@ -3459,6 +3471,66 @@ mod tests {
...
+			cltv_expiry_delta: (4 << 8) | 1,
+			htlc_minimum_msat: 90_000,
+			htlc_maximum_msat: OptionalField::Absent,
+			fee_base_msat: 0,
+			fee_proportional_millionths: 0,
+			excess_data: Vec::new()
+		});
+
+		{
+			// Now, attempt to route 125 sats (just a bit below the capacity of 3 channels).
```
Comment (naumenkogs) [inline lightning/src/routing/router.rs:3659]: outdated

**Comment by naumenkogs** [comment]
Minor comments, ACK 68b62c0116206debe268eca898d48c786d6f16d4

File: lightning/src/routing/router.rs [hunk 6]
```diff
@@ -861,19 +910,29 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+		// because we deterministically terminated the search due to low liquidity.
 		if already_collected_value_msat >= recommended_value_msat || !found_new_path {
 			break 'paths_collection;
 		}
+		// Further, if this was our first walk of the graph, and we weren't limited by an
+		// htlc_minim_msat, return immediately because this path should suffice. If we were limited
+		// by an htlc_minimum_msat value, find another path with a higher value, potentially
+		// allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
+		if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
+			if !hit_minimum_limit {
```
Comment (ariard) [inline lightning/src/routing/router.rs:931]: Following tweak doesn't break router unit tests. It might be covered by new test `exact_fee_liquidity_limit()` but our overpaid value reduction logic is making the outcomes similar ?

```
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 7f00ccdf..e555575b 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -928,9 +928,9 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
                // by an htlc_minimum_msat value, find another path with a higher value, potentially
                // allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
                if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
-                       if !hit_minimum_limit {
-                               break 'paths_collection;
-                       }
+                       //if !hit_minimum_limit {
+                       //      break 'paths_collection;
+                       //}
                        path_value_msat = recommended_value_msat;
                }
        }
```

File: lightning/src/routing/router.rs [hunk 7]
```diff
@@ -143,12 +143,16 @@ struct RouteGraphNode {
 	// - how much is needed for a path being constructed
 	// - how much value can channels following this node (up to the destination) can contribute,
 	//   considering their capacity and fees
-	value_contribution_msat: u64
+	value_contribution_msat: u64,
+	/// The maximum htlc_minimum_msat along the path, taking into consideration the fees required
+	/// to meet the minimum over the hops required to get there.
```
Comment (ariard) [inline lightning/src/routing/router.rs:148]: I think this subtlety around `htlc_minimum_msat` + fees to cover such minimum should be explained better. E.g "The *effective* htlc_minimum_msat required by a hop to forward a payment must also includes fees. `effective_htlc_minimum_msat` = `htlc_minimum_msat` + `fee_base_msat` + (`htlc_minimum_msat` * `fee_proportional_millionths`)

File: lightning/src/routing/router.rs (hunk 4 above)
Comment (ariard) [reply lightning/src/routing/router.rs:789]: You mean the variable renaming or the checking along each hop path that the amount transferred is always above worst-hop's-htlc_minimum_msat-along-the-path ? Where such constraint only apply to hops after such value (i.e closer to payer).

**Comment by TheBlueMatt** [comment]
Will respond to comments soon, but there are further issues I need to debug.

File: lightning/src/routing/router.rs (hunk 4 above)
Comment (naumenkogs) [reply lightning/src/routing/router.rs:789]: I don't understand the renaming.

**Comment by TheBlueMatt** [comment]
Punting to 0.0.14, which we can hopefully accelerate.

File: lightning/src/routing/router.rs [hunk 8]
```diff
@@ -593,20 +650,12 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
-							new_cost = increased_new_cost;
-						} else {
-							new_cost = u64::max_value();
-						}
+						// Ideally the scoring could be smarter (e.g. 0.5*htlc_minimum_msat here),
+						// but it may require additional tracking - we don't want to double-count
+						// the fees included in $incl_fee_next_hops_htlc_minimum_msat, but also
+						// can't use something that may decrease on future hops.
+						let old_cost = cmp::max(old_entry.total_fee_msat, old_entry.path_htlc_minimum_msat);
+						let new_cost = cmp::max(total_fee_msat, $incl_fee_next_hops_htlc_minimum_msat);
```
Comment (TheBlueMatt) [inline lightning/src/routing/router.rs:658]: This is buggy - `old_entry.path_htlc_minimum_msat` is the `path_htlc_minimum_msat` to the same node on the previous iteration, so needs to match here. probably use the path_htlc instead, plus below.
```suggestion
						let new_cost = cmp::max(total_fee_msat, path_htlc_minimum_msat);
```

**Comment by TheBlueMatt** [comment]
Addressed all the bugs I came across, plus some freebie performance tweaks. Still need to address the above comments and add more comments in the changes + write one new test.

File: lightning/src/routing/router.rs (hunk 4 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:840]: Honestly, I'm not sure either. I reverted the change.

File: fuzz/src/router.rs [hunk 9]
```diff
@@ -159,9 +167,11 @@ pub fn do_test<Out: test_logger::Output>(data: &[u8], out: Out) {
 	let our_pubkey = get_pubkey!();
 	let mut net_graph = NetworkGraph::new(genesis_block(Network::Bitcoin).header.block_hash());
 
-	let mut node_pks = HashSet::new();
+	let mut node_pks: HashSet<_, NonRandomHash> = HashSet::default();
 	let mut scid = 42;
 
+	let mut channel_limits = HashMap::new();
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply fuzz/src/router.rs]: Apparently this isn't in the diff anymore? Not sure what it was a reference to.

File: lightning/src/routing/router.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:1003]: I believe that line is a strict performance improvement - if we get there the path we found definitely should be the best one, we could do more work, but it won't find a better path.

**Comment by TheBlueMatt** [comment]
Ok, only a month later, I believe I've addressed all outstanding feedback on this PR, addressed all the issues I've found, and have at least some marginal tests to detect any regressions in the specific issues I've fixed.

File: lightning/src/routing/router.rs [hunk 10]
```diff
@@ -515,18 +579,27 @@
...
+						hit_minimum_limit = true;
+					} else if contributes_sufficient_value {
 						// Note that low contribution here (limited by available_liquidity_msat)
 						// might violate htlc_minimum_msat on the hops which are next along the
 						// payment path (upstream to the payee). To avoid that, we recompute path
 						// path fees knowing the final path contribution after constructing it.
+						let path_htlc_minimum_msat = match compute_fees($incl_fee_next_hops_htlc_minimum_msat, $directional_info.fees)
+								.map(|fee_msat| fee_msat.checked_add($incl_fee_next_hops_htlc_minimum_msat)) {
+							Some(Some(value_msat)) => cmp::max(value_msat, $directional_info.htlc_minimum_msat),
+							_ => u64::max_value()
```
Comment (arik-so) [inline lightning/src/routing/router.rs:601]: if there is no minimum value, why is the default assumption max_u64, as opposed to 0? Does that imply some hops along the way are misconfigured and hence the route must not be used?

File: lightning/src/routing/router.rs [hunk 11]
```diff
@@ -558,120 +628,152 @@
...
-										// max_value means we'll always fail
-										// the old_entry.total_fee_msat > total_fee_msat check
-										total_fee_msat = u64::max_value();
+						#[allow(unused_mut)] // We only use the mut in cfg(test)
+						let mut should_process = !old_entry.was_processed;
+						#[cfg(any(test, feature = "fuzztarget"))]
+						{
+							// In test/fuzzing builds, we do extra checks to make sure the skipping
+							// of already-seen nodes only happens in cases we expect (see below).
+							if !should_process { should_process = true; }
```
Comment (arik-so) [inline lightning/src/routing/router.rs:644]: why wrap it in an if clause?

File: lightning/src/routing/router.rs (hunk 10 above)
Comment (arik-so) [reply lightning/src/routing/router.rs:601]: Never mind, it's checked_add 🙃

File: lightning/src/routing/router.rs [hunk 12]
```diff
@@ -414,25 +459,44 @@
...
 		}
 		if first_hop_targets.is_empty() {
 			return Err(LightningError{err: "Cannot route when there are no outbound routes away from us".to_owned(), action: ErrorAction::IgnoreError});
 		}
 	}
 
+	let empty_channel_features = ChannelFeatures::empty();
+
+	let mut targets = BinaryHeap::new(); //TODO: Do we care about switching to eg Fibbonaci heap?
+	let mut dist = HashMap::with_capacity(network.get_nodes().len());
```
Comment (arik-so) [inline lightning/src/routing/router.rs:480]: I think comments for targets and dist would be very appropriate

File: lightning/src/routing/router.rs [hunk 13]
```diff
@@ -558,120 +628,152 @@
...
 	// meaning how much will be paid in fees after this node (to the best of our knowledge).
 	// This data can later be helpful to optimize routing (pay lower fees).
 	macro_rules! add_entries_to_cheapest_to_target_node {
-		( $node: expr, $node_id: expr, $fee_to_target_msat: expr, $next_hops_value_contribution: expr ) => {
-			if first_hops.is_some() {
-				if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat)) = first_hop_targets.get(&$node_id) {
-					add_entry!(first_hop, *our_node_id, $node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features.to_context(), $fee_to_target_msat, $next_hops_value_contribution);
+		( $node: expr, $node_id: expr, $fee_to_target_msat: expr, $next_hops_value_contribution: expr, $incl_fee_next_hops_htlc_minimum_msat: expr ) => {
+			let skip_node = if let Some(elem) = dist.get_mut($node_id) {
+				let v = elem.was_processed;
```
Comment (arik-so) [inline lightning/src/routing/router.rs:753]: `already_processed` or `had_been_processed` might be a bit better than `v`

File: lightning/src/routing/router.rs [hunk 14]
```diff
@@ -558,120 +628,152 @@
...
+			if !skip_node && first_hops.is_some() {
+				if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&$node_id) {
+					add_entry!(first_hop, *our_node_id, $node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, $fee_to_target_msat, $next_hops_value_contribution, $incl_fee_next_hops_htlc_minimum_msat);
 				}
 			}
 
-			let features;
-			if let Some(node_info) = $node.announcement_info.as_ref() {
-				features = node_info.features.clone();
+			let features = if let Some(node_info) = $node.announcement_info.as_ref() {
```
Comment (arik-so) [inline lightning/src/routing/router.rs:770]: afaict, features is only used if skip_node is false? You might be able to combine the skip_node checks from the if clauses above and below and only access features inside it

File: lightning/src/routing/router.rs [hunk 15]
```diff
@@ -851,18 +954,29 @@
...
 					*channel_liquidity_available_msat -= spent_on_hop_msat;
 				}
+				if !prevented_redundant_path_selection {
+					// If we weren't capped by hitting a liquidity limit on a channel in the path,
+					// we'll probably end up picking the same path again on the next iteration.
+					// Randomly decrease the available liquidity of a hop in the middle of the
+					// path.
+					let victim_liquidity = bookkeeped_channels_liquidity_available_msat.get_mut(
+						&payment_path.hops[(payment_path.hops.len() - 1) / 2].0.short_channel_id).unwrap();
+					*victim_liquidity = 0;
```
Comment (arik-so) [inline lightning/src/routing/router.rs:977]: lol at the name

File: lightning/src/routing/router.rs [hunk 16]
```diff
@@ -3400,6 +3538,361 @@
 			assert_eq!(total_amount_paid_msat, 90_000);
 		}
 	}
+
+	#[test]
+	fn min_criteria_consistency() {
+		// Test that we don't use an inconsistent metric between updating and walking nodes during
+		// our Dijkstra's pass. In the initial version of MPP, the "best source" for a given node
+		// was updated with a different critera from the heap sorting, resulting in loops in
```
Comment (arik-so) [inline lightning/src/routing/router.rs:3546]: nit: a different criterion

File: lightning/src/routing/router.rs [hunk 17]
```diff
@@ -3400,6 +3538,361 @@
...
+		// calculated paths. We test for that specific case here.
+
+		// We construct a network that looks like this:
+		//
+		//            node2 -1(3)2- node3
```
Comment (arik-so) [inline lightning/src/routing/router.rs:3551]: is the bracketed number the value locked up in HTLCs?

File: lightning/src/routing/router.rs [hunk 18]
```diff
@@ -193,20 +202,33 @@
...
+	value_contribution_msat: u64,
 }
 
 // Instantiated with a list of hops with correct data in them collected during path finding,
 // an instance of this struct should be further modified only via given methods.
 #[derive(Clone)]
-struct PaymentPath {
-	hops: Vec<PathBuildingHop>,
+struct PaymentPath<'a> {
+	hops: Vec<(PathBuildingHop<'a>, NodeFeatures)>,
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:225]: PathBuildingHop is created for each node as we walk the graph, but PaymentPath only at the end. By avoiding keeping Features (which are Vecs under the hood) in the graph map during the main graph walk, we avoid a lot of Vec clones.

File: lightning/src/routing/router.rs [hunk 19]
```diff
@@ -414,25 +459,44 @@
...
+	const ROUTE_CAPACITY_PROVISION_FACTOR: u64 = 3;
+	let recommended_value_msat = final_value_msat * ROUTE_CAPACITY_PROVISION_FACTOR as u64;
+	let mut path_value_msat = final_value_msat;
+
 	// We don't want multiple paths (as per MPP) share liquidity of the same channels.
 	// This map allows paths to be aware of the channel use by other paths in the same call.
 	// This would help to make a better path finding decisions and not "overbook" channels.
 	// It is unaware of the directions (except for `outbound_capacity_msat` in `first_hops`).
-	let mut bookkeeped_channels_liquidity_available_msat = HashMap::new();
+	let mut bookkeeped_channels_liquidity_available_msat = HashMap::with_capacity(network.get_nodes().len());
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:499]: Heh, not my variable name, in transliterated english it makes sense, though :p.

File: lightning/src/routing/router.rs (hunk 11 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:644]: Because otherwise rust complains that the initial assignment is never read from.

File: lightning/src/routing/router.rs (hunk 14 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:770]: Yea, doesn't matter much, but, sure.

File: lightning/src/routing/router.rs (hunk 17 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:3556]: Its the short_channel_id. I think this is documented or at least more clear in the graphs in some of the earlier tests.

File: lightning/src/routing/router.rs [hunk 20]
```diff
@@ -193,20 +202,33 @@ struct PathBuildingHop {
...
+	/// If we've already processed a node as the best node, we shouldn't process it again. Normally
+	/// we'd just ignore it if we did as all channels would have a higher new fee, but because we
+	/// may decrease the amounts in use as we walk the graph, the actual calculated fee may
+	/// decrease as well. Thus, we have to explicitly track which nodes have been processed and
+	/// avoid processing them again.
+	was_processed: bool,
+	#[cfg(test)]
+	// In tests, we apply further sanity checks on cases where we skip nodes we already processed
+	// to ensure it is specifically in cases where the fee has gone down because of a decrease in
+	// valud_contribution_msat, which requires tracking it here.
```
Comment (ariard) [inline lightning/src/routing/router.rs:217]: "we skip nodes we already processed to ensure it is specifically" which requirement is ensured ? Also "valued_contribution_msat"

File: lightning/src/routing/router.rs [hunk 21]
```diff
@@ -226,16 +248,14 @@ impl PaymentPath {
...
-	// (in terms of reaching htlc_minimum_msat), so that this function puts the fees in order.
-	// In that case we call it on the "same" amount we initially allocated for this path, and which
-	// could have been reduced on the way. In that case, there is also a risk of exceeding
-	// available_liquidity inside this function, because the function is unaware of this bound.
-	// In our specific recomputation cases where we never increase the value the risk is pretty low.
-	// This function, however, does not support arbitrarily increasing the value being transferred,
-	// and the exception will be triggered.
+	// Sometimes we call this function right after constructing a path which is inconsistent in
+	// that it the value being transferred has decreased while we were doing path finding, leading
+	// to the fees being paid not linig up with the actual limits.
```
Comment (ariard) [inline lightning/src/routing/router.rs:253]: nit: "lining up with the actual limits"

File: lightning/src/routing/router.rs [hunk 22]
```diff
@@ -372,8 +392,42 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+	// sent along the path for any hops after that channel. This may imply that later fees (which
+	// we've already tabulated) are lower because a smaller value is passing through the channels
+	// (and the proportional fee is thus lower). There isn't a trivial way to recalculate the
+	// channels which were selected earlier (and which may still be used for other paths without a
+	// lower liquidity limit), so we simply accept that some liquidity-limited paths may be
+	// de-preferenced.
+	//
+	// One potentially problematic case for this algorithm would be if there are many
+	// liquidity-limited paths which are liquidity-limited near the destination (ie early in our
+	// graph walking), we may never find a liquidity-unlimited path which has lower proportional
```
Comment (ariard) [inline lightning/src/routing/router.rs:411]: what's a liquidity-unlimited path ? I think it's comparative there and not binary ?

**Review by ariard — COMMENTED** [review]
Ongoing review, just few nits

File: lightning/src/routing/router.rs (hunk 22 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:411]: Canged to "a path which is not liquidity-limited".

**Review by arik-so — APPROVED** [review]

File: lightning/src/routing/router.rs [hunk 23]
```diff
@@ -558,140 +634,173 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+										hop_use_fee_msat = fee_msat;
+										total_fee_msat += hop_use_fee_msat;
+										// When calculating the lowest inbound fees to a node, we
+										// calculate fees here not based on the actual value we think
+										// will flow over this channel, but on the minimum value that
+										// we'll accept flowing over it. Otherwise we may later find a
+										// different path to the source node that is more expensive,
+										// but which we consider to be cheaper because we are capacity
+										// constrained and the relative fee becomes lower.
+										match compute_fees(minimal_value_contribution_msat, old_entry.src_lowest_inbound_fees)
```
Comment (ariard) [inline lightning/src/routing/router.rs:674]: Maybe comment can be clearer "...but on the minimum value that we'll accept flowing over it. Minimal value contribution being a constant, the compared fees from different path to a source node will have the same basis. Thus avoiding to bias toward a capacity-constrained path with a higher relative fee" ?

**Review by ariard — COMMENTED** [review]
Ongoing review so far up to dc01b0e

File: lightning/src/routing/router.rs [hunk 24]
```diff
@@ -372,8 +392,42 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 	// 8. Choose the best route by the lowest total fee.
 
 	// As for the actual search algorithm,
-	// we do a payee-to-payer Dijkstra's sorting by each node's distance from the payee
-	// plus the minimum per-HTLC fee to get from it to another node (aka "shitty A*").
+	// we do a payee-to-payer pseudo-Dijkstra's sorting by each node's distance from the payee
+	// plus the minimum per-HTLC fee to get from it to another node (aka "shitty pseudo-A*").
+	//
+	// We are not a faithful Dijkstra's implementation because we can change values which impact
+	// earlier nodes while processing later nodes. Specifically, if we reach a channel with a lower
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:400]: Hmm, I read it as "the liquidity limit is lower" not "the lowest liquidity value". I could rephrase to "lower liquidity upper-bound"?

File: lightning/src/routing/router.rs [hunk 25]
```diff
@@ -372,8 +392,42 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+	// lower liquidity limit), so we simply accept that some liquidity-limited paths may be
+	// de-preferenced.
+	//
+	// One potentially problematic case for this algorithm would be if there are many
+	// liquidity-limited paths which are liquidity-limited near the destination (ie early in our
+	// graph walking), we may never find a liquidity-unlimited path which has lower proportional
+	// fee (and only lower absolute fee when considering the ultimate value sent). Because we only
+	// consider paths with at least 5% of the total value being sent, the damage from such a case
+	// should be limited, however this could be further reduced in the future by calculating fees
+	// on the amount we wish to route over a path, not the amount we are routing over a path.
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs]: I changed the sentence to end with "ie ignoring the liquidity limits for the purposes of fee calculation."

File: lightning/src/routing/router.rs [hunk 26]
```diff
@@ -851,18 +961,29 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+						// If this path used all of this channel's available liquidity, we know
+						// this path will not be selected again in the next loop iteration.
+						prevented_redundant_path_selection = true;
 					}
 					*channel_liquidity_available_msat -= spent_on_hop_msat;
 				}
+				if !prevented_redundant_path_selection {
+					// If we weren't capped by hitting a liquidity limit on a channel in the path,
+					// we'll probably end up picking the same path again on the next iteration.
+					// Randomly decrease the available liquidity of a hop in the middle of the
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:980]: The issue here is you may be sending 1/1000th of the value of the node in the middle. Without this commit, we just select the same route over and over again and never find any other options. Eventually we should do something somewhat smarter, like checking if a route is exactly identical to one we already selected, but just avoiding the same middle hop solves the issue as well for now.

I dropped the "random", though it was a reference to the somewhat random decision to do this.

File: lightning/src/routing/router.rs [hunk 27]
```diff
@@ -414,25 +459,50 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+	const ROUTE_CAPACITY_PROVISION_FACTOR: u64 = 3;
+	let recommended_value_msat = final_value_msat * ROUTE_CAPACITY_PROVISION_FACTOR as u64;
+	let mut path_value_msat = final_value_msat;
+
 	// We don't want multiple paths (as per MPP) share liquidity of the same channels.
 	// This map allows paths to be aware of the channel use by other paths in the same call.
 	// This would help to make a better path finding decisions and not "overbook" channels.
 	// It is unaware of the directions (except for `outbound_capacity_msat` in `first_hops`).
-	let mut bookkeeped_channels_liquidity_available_msat = HashMap::new();
+	let mut bookkeeped_channels_liquidity_available_msat = HashMap::with_capacity(network.get_nodes().len());
```
Comment (ariard) [inline lightning/src/routing/router.rs:505]: Pre-allocating to the size of the whole graph assumes we'll completely traverse it before reaching `recommended_value_msat` or `path_value_msat` liquidity threshold ? Or what's your thinking there ?

**Review by ariard — COMMENTED** [review]
I think aa9c2c2 is worthy of better explanations. Otherwise, it sounds good to me modulo squashing.

File: lightning/src/routing/router.rs (hunk 27 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:505]: Yea, I mean I guess we could divide by some factor, but why bother, its short-lived and at worst we will, so we might as well just allocate for it.

File: lightning/src/routing/router.rs [hunk 28]
```diff
@@ -558,140 +634,175 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+							} else if old_entry.was_processed && new_cost < old_cost {
+								#[cfg(test)]
+								{
+									// If we're skipping processing a node which was previously
+									// processed even though we found another path to it with a
+									// cheaper fee, check that it was because the second path we
+									// found (which we are processing now) has a lower value
+									// contribution due to an HTLC minimum limit.
+									debug_assert!(path_htlc_minimum_msat < old_entry.path_htlc_minimum_msat);
+									debug_assert!(value_contribution_msat < old_entry.value_contribution_msat);
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:756]: I added a note (see diff below) and squashed.

**Comment by TheBlueMatt** [comment]
Expanded comment and included some extra fuzzing tests. Diff is
```
$ git diff-tree -U2 1bde367 HEAD
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 68815ff2..a74f0cf5 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -212,5 +212,5 @@ struct PathBuildingHop<'a> {
 	/// avoid processing them again.
 	was_processed: bool,
-	#[cfg(test)]
+	#[cfg(any(test, feature = "fuzztarget"))]
 	// In tests, we apply further sanity checks on cases where we skip nodes we already processed
 	// to ensure it is specifically in cases where the fee has gone down because of a decrease in
@@ -637,5 +637,5 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 								path_htlc_minimum_msat,
 								was_processed: false,
-								#[cfg(test)]
+								#[cfg(any(test, feature = "fuzztarget"))]
 								value_contribution_msat,
 							}
@@ -727,10 +727,10 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 								old_entry.htlc_minimum_msat = $directional_info.htlc_minimum_msat;
 								old_entry.path_htlc_minimum_msat = path_htlc_minimum_msat;
-								#[cfg(test)]
+								#[cfg(any(test, feature = "fuzztarget"))]
 								{
 									old_entry.value_contribution_msat = value_contribution_msat;
 								}
 							} else if old_entry.was_processed && new_cost < old_cost {
-								#[cfg(test)]
+								#[cfg(any(test, feature = "fuzztarget"))]
 								{
 									// If we're skipping processing a node which was previously
@@ -739,4 +739,18 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 									// found (which we are processing now) has a lower value
 									// contribution due to an HTLC minimum limit.
+									//
+									// e.g. take a graph with two paths from node 1 to node 2, one
+									// through channel A, and one through channel B. Channel A and
+									// B are both in the to-process heap, with their scores set by
+									// a higher htlc_minimum than fee.
+									// Channel A is processed first, and the channels onwards from
+									// node 1 are added to the to-process heap. Thereafter, we pop
+									// Channel B off of the heap, note that it has a much more
+									// restrictive htlc_maximum_msat, and recalculate the fees for
+									// all of node 1's channels using the new, reduced, amount.
+									//
+									// This would be bogus - we'd be selecting a higher-fee path
+									// with a lower htlc_maximum_msat instead of the one we'd
+									// already decided to use.
 									debug_assert!(path_htlc_minimum_msat < old_entry.path_htlc_minimum_msat);
 									debug_assert!(value_contribution_msat < old_entry.value_contribution_msat);
```

File: lightning/src/routing/router.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/routing/router.rs:400]: Yes I like "lower liquidity upper-bound".

File: lightning/src/routing/router.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/routing/router.rs]: Yeah good to me for now. I don't have a better mitigation against repeated path selection to propose rn.

File: lightning/src/routing/router.rs [hunk 29]
```diff
@@ -558,140 +634,189 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
...
+									// contribution due to an HTLC minimum limit.
+									//
+									// e.g. take a graph with two paths from node 1 to node 2, one
+									// through channel A, and one through channel B. Channel A and
+									// B are both in the to-process heap, with their scores set by
+									// a higher htlc_minimum than fee.
+									// Channel A is processed first, and the channels onwards from
+									// node 1 are added to the to-process heap. Thereafter, we pop
+									// Channel B off of the heap, note that it has a much more
+									// restrictive htlc_maximum_msat, and recalculate the fees for
```
Comment (ariard) [inline lightning/src/routing/router.rs:749]: Straightforward comment thanks! Should this say "htlc_minimum" and under ?

**Review by ariard — COMMENTED** [review]
Code Review ACK 5f5bc41 modulo nit

File: lightning/src/routing/router.rs (hunk 29 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:749]: Hmm, I dont think so, the issue appears when the `value_contribution_msat` is lower (which is often just the path `htlc_maximum_msat`). The `htlc_minimum_msat` check here is about the criteria for when it should be possible to hit this case.

**Comment by TheBlueMatt** [comment]
Gonna merge cause I'm tired of CI failing. I promise to open a new PR to fix the comment if I'm wrong :).
